pkg/scm/git/git.go

```diff
@@ -62,18 +68,103 @@ func stringInSlice(s string, slice []string) bool {
 // ValidCloneSpec determines if the given string reference points to a valid git
 // repository
 func (h *stiGit) ValidCloneSpec(source string) bool {
-	url, err := url.Parse(source)
-	if err != nil {
+	return validCloneSpec(source, false, nil)
+}
+
+// ValidCloneSpecRemoteOnly determines if the given string reference points to a valid remote git
+// repository; valid local repository specs will result in a return of false
+func (h *stiGit) ValidCloneSpecRemoteOnly(source string) bool {
+	return validCloneSpec(source, true, nil)
+}
+
+// MungeNoProtocolURL will take a URL returned from net/url.Parse and make
+// corrections to the URL when no protocol is specified in the Scheme
+func (h *stiGit) MungeNoProtocolURL(source string, url *url.URL) bool {
+	return validCloneSpec(source, false, url)
+}
+
+func validCloneSpec(source string, remoteOnly bool, partialURL *url.URL) bool {
+	if partialURL == nil {
+		uri, err := url.Parse(source)
+		if err != nil {
+			return false
+		}
+		partialURL = uri
+	}
+
+	if stringInSlice(partialURL.Scheme, allowedSchemes) {
+		return true
+	}
+
+	if partialURL.Scheme != "" {
 		return false
 	}
 
-	if stringInSlice(url.Scheme, allowedSchemes) {
+	user := ""
+	host := ""
+	path := ""
+	ref := ""
+	segments := strings.SplitN(source, "@", 2)
+	if len(segments) == 2 {
+		user = segments[0]
+		if !gitSshURLUser.MatchString(user) {
+			return false
+		}
+
+		host = segments[1]
+		splitChar := ":"
+		if gitSshURLIPVersionCheck.MatchString(host) {
+			splitChar = "/"
+		}
+
+		segments = strings.SplitN(host, splitChar, 2)
+		if len(segments) == 2 {
+			host = segments[0]
+			if !gitSshURLIPv6.MatchString(host) && !gitSshURLIPv4.MatchString(host) {
+				return false
+			}
+
+			path = segments[1]
+			segments = strings.SplitN(path, "#", 2)
+			if len(segments) == 2 {
+				path, ref = segments[0], segments[1]
+
+				if !gitSshURLPathRef.MatchString(ref) {
+					return false
+				}
+			}
+
+			if !gitSshURLPathRef.MatchString(path) {
+				return false
+			}
+		}
+		partialURL.Scheme = "ssh"
+		partialURL.User = url.User(user)
+		partialURL.Host = host
+		partialURL.Path = path
+		partialURL.Fragment = ref
+		return true
+	} else if !remoteOnly {
```

Comment:
If the user string is empty, we end up with the same empty Username struct that url.Parse(string) started us on.  My take was that a simple re-initialization was cleaner that and if len(user) > 0 check.

Based on that clarification, would you agree?